issue: Conditional/dynamic schema resolution no longer working

### Version Number

5.1.1

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/angry-butterfly-ljt6z2

### Steps to reproduce

Please see code sandbox. I get the error when when I conditionally set the `formResolver`

### Expected behaviour

Previously, I was able to conditionally pass the schema to denote if it was update or create based on others props (please see code below and in code sandbox). However, since upgrading to 5.1.1 from 5.0.1, zodResolver gives me type errors. 

```ts
export function AgencyForm({ agency }: { agency?: UpdateAgencyType }) {
  const defaultValues = agency
    ? { ...agency, legalName: agency.legalName ?? "" }
    : {
        slug: "",
        name: "",
        legalName: "",
      };
  const formResolver = agency ? updateAgencySchema : createAgencySchema;
  const form = useForm<z.infer<typeof formResolver>>({
    mode: "onBlur",
    reValidateMode: "onBlur",
    resolver: zodResolver(
      formResolver as z.ZodType<z.infer<typeof formResolver>>
    ),
    defaultValues,
  });
```

Here's the error I"m getting: 

```
No overload matches this call.
  Overload 1 of 4, '(schema: Zod3Type<{ slug: string; name: string; acronym: string; legalName: string | null; } | { cuid: string; slug?: string | undefined; name?: string | undefined; acronym?: string | undefined; legalName?: string | ... 1 more ... | undefined; }, { ...; } | { ...; }>, schemaOptions?: ParseParams | undefined, resolverOptions?: NonRawResolverOptions | undefined): Resolver<...>', gave the following error.
    Argument of type 'ZodType<{ slug: string; name: string; acronym: string; legalName: string | null; } | { cuid: string; slug?: string | undefined; name?: string | undefined; acronym?: string | undefined; legalName?: string | ... 1 more ... | undefined; }, ZodTypeDef, { ...; } | { ...; }>' is not assignable to parameter of type 'Zod3Type<{ slug: string; name: string; acronym: string; legalName: string | null; } | { cuid: string; slug?: string | undefined; name?: string | undefined; acronym?: string | undefined; legalName?: string | ... 1 more ... | undefined; }, { ...; } | { ...; }>'.
      Types of property '_def' are incompatible.
        Property 'typeName' is missing in type 'ZodTypeDef' but required in type '{ typeName: string; }'.
  Overload 2 of 4, '(schema: $ZodType<unknown, FieldValues, $ZodTypeInternals<unknown, FieldValues>>, schemaOptions?: ParseContext<$ZodIssue> | undefined, resolverOptions?: NonRawResolverOptions | undefined): Resolver<...>', gave the following error.
    Argument of type 'ZodType<{ slug: string; name: string; acronym: string; legalName: string | null; } | { cuid: string; slug?: string | undefined; name?: string | undefined; acronym?: string | undefined; legalName?: string | ... 1 more ... | undefined; }, ZodTypeDef, { ...; } | { ...; }>' is not assignable to parameter of type '$ZodType<unknown, FieldValues, $ZodTypeInternals<unknown, FieldValues>>'.
      Property '_zod' is missing in type 'ZodType<{ slug: string; name: string; acronym: string; legalName: string | null; } | { cuid: string; slug?: string | undefined; name?: string | undefined; acronym?: string | undefined; legalName?: string | ... 1 more ... | undefined; }, ZodTypeDef, { ...; } | { ...; }>' but required in type '$ZodType<unknown, FieldValues, $ZodTypeInternals<unknown, FieldValues>>'.

```

### What browsers are you seeing the problem on?

_No response_

### Relevant log output

```shell

```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

Having the same issue when upgrading from 5.0.1 to 5.1.1. My use case is for doing i18n in zod schema messages.

```typescript
import{ type TFunction } from 'i18next';
import { zodResolver } from '@hookform/resolvers/zod';
import { useEffect, useMemo } from 'react';
import { useForm, type FieldValues, type UseFormProps, type UseFormReturn } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import type { ZodType, ZodTypeDef } from 'zod';

type I18nNamespaces = 'namespace1' | 'namespace2';
type TFunction = TFunction<I18nNamespaces, undefined>;

type SchemaBuilder<TFormInput, TFormOutput> = (t: TFunction) => ZodType<TFormOutput, ZodTypeDef, TFormInput>;

// Replicating react-hook-form useForm signature generics (including order)
export function useI18nForm<
  TFormInput extends FieldValues,
  TFormContext = unknown,
  TFormOutput extends FieldValues = TFormInput
>(
  schemaBuilder: SchemaBuilder<TFormInput, TFormOutput>,
  formProps: Omit<UseFormProps<TFormInput, TFormContext, TFormOutput>, 'resolver'>,
  t: TFunction
): UseFormReturn<TFormInput, TFormContext, TFormOutput> {
  const { i18n } = useTranslation();

  const schema = useMemo(() => schemaBuilder(t), [schemaBuilder, t]);

  const methods = useForm<TFormInput, TFormContext, TFormOutput>({
    ...formProps,
    resolver: zodResolver(schema)
  });

  const reset = methods.reset;

  useEffect(() => {
    void i18n.language;
    reset(undefined, { keepValues: true });
  }, [i18n.language, reset]);

  return methods;
}
```

Type error:

```
error TS2769: No overload matches this call.
  Overload 1 of 4, '(schema: Zod3Type<TFormOutput, TFormInput>, schemaOptions?: ParseParams | undefined, resolverOptions?: NonRawResolverOptions | undefined): Resolver<...>', gave the following error.
    Argument of type 'ZodType<TFormOutput, ZodTypeDef, TFormInput>' is not assignable to parameter of type 'Zod3Type<TFormOutput, TFormInput>'.
      Types of property '_def' are incompatible.
        Property 'typeName' is missing in type 'ZodTypeDef' but required in type '{ typeName: string; }'.
  Overload 2 of 4, '(schema: $ZodType<unknown, FieldValues>, schemaOptions?: ParseContext<$ZodIssue> | undefined, resolverOptions?: NonRawResolverOptions | undefined): Resolver<...>', gave the following error.
    Argument of type 'ZodType<TFormOutput, ZodTypeDef, TFormInput>' is not assignable to parameter of type '$ZodType<unknown, FieldValues>'.
      Property '_zod' is missing in type 'ZodType<TFormOutput, ZodTypeDef, TFormInput>' but required in type '$ZodType<unknown, FieldValues>'.
```

--- a/zod/src/__tests__/zod-v3.ts
+++ b/zod/src/__tests__/zod-v3.ts
@@ -214,4 +214,27 @@ describe('zodResolver', () => {
       }>
     >();
   });
+
+  it('should accept a dynamically selected schema typed via the generic z.ZodType annotation', async () => {
+    // https://github.com/react-hook-form/resolvers/issues/782 — schemas
+    // chosen conditionally between two shapes and annotated with the base
+    // `z.ZodType<Output>` (rather than a concrete subclass like `ZodObject`)
+    // don't statically expose `_def.typeName`, even though they're genuine
+    // Zod 3 schema instances at runtime.
+    const updateSchema = z.object({ id: z.string(), name: z.string() });
+    const createSchema = z.object({ name: z.string() });
+    const useUpdateSchema = true;
+    const dynamicSchema = (
+      useUpdateSchema ? updateSchema : createSchema
+    ) as z.ZodType<{ name: string }>;
+
+    const resolver = zodResolver(dynamicSchema);
+
+    const result = await resolver({ id: '1', name: 'John' } as any, undefined, {
+      fields: {},
+      shouldUseNativeValidation,
+    });
+
+    expect(result.errors).toEqual({});
+  });
 });

--- a/zod/src/__tests__/zod-v4.ts
+++ b/zod/src/__tests__/zod-v4.ts
@@ -208,6 +208,30 @@ describe('zodResolver', () => {
     });
   });
 
+  it('should not misdetect a zod4 schema as zod3 (both have an object-shaped `_def`)', async () => {
+    // The zod3 detection was loosened to only require an object-shaped
+    // `_def` (see issue #782 — dynamically selected schemas typed via the
+    // generic `z.ZodType<Output>` annotation don't statically expose
+    // `_def.typeName`). Zod 4 schemas also have an object-shaped `_def`
+    // (just without `typeName`), so zod4 must be detected first via its
+    // unambiguous `_zod` marker — otherwise this would incorrectly take the
+    // zod3 code path and mishandle zod4's error shape.
+    const testSchema = z.object({ name: z.string() });
+
+    const result = await zodResolver(testSchema)(
+      { name: 123 as any },
+      undefined,
+      {
+        fields: {},
+        shouldUseNativeValidation,
+      },
+    );
+
+    expect(result.errors).toMatchObject({
+      name: { type: 'invalid_type' },
+    });
+  });
+
   it('should throw any error unrelated to Zod', async () => {
     const schemaWithCustomError = schema.refine(() => {
       throw Error('custom error');

--- a/zod/src/zod.ts
+++ b/zod/src/zod.ts
@@ -14,11 +14,11 @@ const isZod3Error = (error: any): error is z3.ZodError => {
   return Array.isArray(error?.issues);
 };
 const isZod3Schema = (schema: any): schema is z3.ZodSchema => {
-  return (
-    '_def' in schema &&
-    typeof schema._def === 'object' &&
-    'typeName' in schema._def
-  );
+  // Deliberately doesn't require `typeName` in `_def` — dynamically selected
+  // schemas typed via the generic `z.ZodType<Output>` annotation (rather than
+  // a concrete subclass like `ZodObject`) don't statically expose `typeName`,
+  // even though they're genuine Zod 3 schema instances at runtime.
+  return '_def' in schema && typeof schema._def === 'object';
 };
 const isZod4Error = (error: any): error is z4.$ZodError => {
   // instanceof is safe in Zod 4 (uses Symbol.hasInstance)
@@ -165,9 +165,7 @@ type NonRawResolverOptions = {
 interface Zod3Type<O = unknown, I = unknown> {
   _output: O;
   _input: I;
-  _def: {
-    typeName: string;
-  };
+  _def: object;
 }
 
 // minimal interface to avoid assignability issues between zod v4 patch/minor
@@ -275,40 +273,11 @@ export function zodResolver<Input extends FieldValues, Context, Output>(
     raw?: boolean;
   } = {},
 ): Resolver<Input, Context, Output | Input> {
-  if (isZod3Schema(schema)) {
-    return async (values: Input, _, options) => {
-      try {
-        const data = await schema[
-          resolverOptions.mode === 'sync' ? 'parse' : 'parseAsync'
-        ](values, schemaOptions);
-
-        options.shouldUseNativeValidation &&
-          validateFieldsNatively({}, options);
-
-        return {
-          errors: {},
-          values: resolverOptions.raw ? Object.assign({}, values) : data,
-        } satisfies ResolverSuccess<Output | Input>;
-      } catch (error) {
-        if (isZod3Error(error)) {
-          return {
-            values: {},
-            errors: toNestErrors(
-              parseZod3Issues(
-                error.errors,
-                !options.shouldUseNativeValidation &&
-                  options.criteriaMode === 'all',
-              ),
-              options,
-            ),
-          } satisfies ResolverError<Input>;
-        }
-
-        throw error;
-      }
-    };
-  }
-
+  // Zod 4 is checked first: its `_zod` marker is unambiguous, whereas the
+  // Zod 3 check below only requires an object-shaped `_def` (to support
+  // dynamically-selected schemas typed via the generic `z.ZodType<Output>`
+  // annotation, see `isZod3Schema`) — and Zod 4 schemas also have an
+  // object-shaped `_def`, so they'd otherwise be misdetected as Zod 3.
   if (isZod4Schema(schema)) {
     return async (values: Input, _, options) => {
       try {
@@ -348,6 +317,40 @@ export function zodResolver<Input extends FieldValues, Context, Output>(
             errors: toNestErrors(
               parseZod4Issues(
                 error.issues,
+                !options.shouldUseNativeValidation &&
+                  options.criteriaMode === 'all',
+              ),
+              options,
+            ),
+          } satisfies ResolverError<Input>;
+        }
+
+        throw error;
+      }
+    };
+  }
+
+  if (isZod3Schema(schema)) {
+    return async (values: Input, _, options) => {
+      try {
+        const data = await schema[
+          resolverOptions.mode === 'sync' ? 'parse' : 'parseAsync'
+        ](values, schemaOptions);
+
+        options.shouldUseNativeValidation &&
+          validateFieldsNatively({}, options);
+
+        return {
+          errors: {},
+          values: resolverOptions.raw ? Object.assign({}, values) : data,
+        } satisfies ResolverSuccess<Output | Input>;
+      } catch (error) {
+        if (isZod3Error(error)) {
+          return {
+            values: {},
+            errors: toNestErrors(
+              parseZod3Issues(
+                error.errors,
                 !options.shouldUseNativeValidation &&
                   options.criteriaMode === 'all',
               ),

